ledger/store/trackerdb/pebbledbdriver/pebbledriver.go

```diff
@@ -0,0 +1,572 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package pebbledbdriver
+
+import (
+	"context"
+	"io"
+	"runtime"
+	"time"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/ledger/store/trackerdb"
+	"github.com/algorand/go-algorand/ledger/store/trackerdb/generickv"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/util/db"
+	"github.com/cockroachdb/pebble"
+	"github.com/cockroachdb/pebble/bloom"
+	"github.com/cockroachdb/pebble/vfs"
+)
+
+const (
+	// minCache is the minimum amount of memory in megabytes to allocate to pebble
+	// read and write caching, split half and half.
+	minCache = 16
+
+	// minHandles is the minimum number of files handles to allocate to the open
+	// database files.
+	minHandles = 16
+)
+
+type trackerStore struct {
+	kvs   kvstore
+	proto config.ConsensusParams
+	// use the generickv implementations
+	trackerdb.Reader
+	trackerdb.Writer
+	trackerdb.Catchpoint
+}
+
+// Open opens a Pebble db database
+func Open(dbdir string, inMem bool, proto config.ConsensusParams, log logging.Logger) (trackerdb.Store, error) {
+	cache := 128 // this divided by 2 and by memTableLimit = 1GB /(2 * 16) = 32MB per memtable
+	handles := 1000
+
+	// Ensure we have some minimal caching and file guarantees
+	if cache < minCache {
+		cache = minCache
+	}
+	if handles < minHandles {
+		handles = minHandles
+	}
+
+	// The max memtable size is limited by the uint32 offsets stored in
+	// internal/arenaskl.node, DeferredBatchOp, and flushableBatchEntry.
+	// Taken from https://github.com/cockroachdb/pebble/blob/master/open.go#L38
+	maxMemTableSize := 4<<30 - 1 // Capped by 4 GB
+
+	memTableLimit := 2 // default: 2
+	memTableSize := cache * 1024 * 1024 / 2 / memTableLimit
+	if memTableSize > maxMemTableSize {
+		memTableSize = maxMemTableSize
+	}
+
+	// configure pebbledb
+	opts := &pebble.Options{
+		// logging
+		Logger: log,
+
+		// Pebble has a single combined cache area and the write
+		// buffers are taken from this too. Assign all available
+		// memory allowance for cache.
+		Cache:        pebble.NewCache(int64(cache * 1024 * 1024)), // default: 8 MB
+		MaxOpenFiles: handles,                                     // default: 1000
```

Comment:
what they are doing with files if a single db wants 1000 open descriptors?
We have limits calculation in server.go so it should be updated accordingly.
```
daemon/algod/server.go: err = util.SetFdSoftLimit(fdRequired)
```

Reply:
pebble has lots of files, is not a single massive file with internal management like some other dbs.
I haven't looked at how much they use when, but my guess is concurrent compaction needing the most handles, particularly on deeper levels.

Reply:
so need to update `server.go` calculation